Rewrite root README (#126)
Shopify/argo: README.md

If you want to learn a little more about the patterns found throughout these libraries, and the way that they are rendered into the application they are extending, read on!

## What is Argo, really?

Argo is the name we give to the patterns we provide for extending our user interfaces. It gives us a short, memorable name to use to refer to the components, APIs, patterns, and the underlying technology that makes these extensions possible. Let’s dig into some of the things that add up to create Argo.

### Open-source core

The underlying technology for Argo is [remote-ui](https://github.com/Shopify/remote-ui), and open source technology built by Shopify. remote-ui provides the basic [message passing](https://github.com/Shopify/remote-ui/tree/main/packages/rpc) system that is used by an Argo extension to communicate with the “host” application it is extending. remote-ui also provides the [component model](https://github.com/Shopify/remote-ui/tree/main/packages/core). If you are familiar with building for the web, remote-ui is very similar to the DOM — it gives you a programmatic model for defining UI components and attaching UI to the screen.

> In addition to the basic message passing and component model, remove-ui offers integrations for frameworks like [React](https://github.com/Shopify/remote-ui/tree/main/packages/react) and [Vue](https://github.com/Shopify/remote-ui/tree/main/packages/vue), which we make use of within Argo to provide framework-specific bindings with little additional effort.

Reviewer: ```suggestion
In addition to the basic message passing and component model, remote-ui offers integrations for frameworks like [React](https://github.com/Shopify/remote-ui/tree/main/packages/react) and [Vue](https://github.com/Shopify/remote-ui/tree/main/packages/vue), which we make use of within Argo to provide framework-specific bindings with little additional effort.
```

The exact components available to the extension depend on the surface area you are embedding — please refer to the [documentation linked above](#i-just-want-to-build-an-argo-extension-not-learn-about-them) for references on what components you can use. If you’ve ever used [Polaris](https://polaris.shopify.com/), though, the components you will find will look extremely familiar, as many components and props are based on their Polaris equivalents.

### Extension points

> While remote-ui provides the component model, and takes care of propagating updates to an extension’s UI into the host application, it does not have any built-in notion of an “extension” at all. To create an extension system, Shopify authors a bit of extra code: a mapping of “extension points” (which, in code, are just strings with a specific naming format), and a way to load third party code that can register to be called for those extension points. This might sound little complicated, but the code to put it all together is actually quite straightforward. The Argo ”runtime” code, which is used to construct the sandbox your extension will run in, looks almost exactly like this for all Argo-capable surface areas in Shopify:

Reviewer 2: typo: This might sound **a** little complicated, ...

> While remote-ui provides the component model, and takes care of propagating updates to an extension’s UI into the host application, it does not have any built-in notion of an “extension” at all. To create an extension system, Shopify authors a bit of extra code: a mapping of “extension points” (which, in code, are just strings with a specific naming format), and a way to load third party code that can register to be called for those extension points. This might sound little complicated, but the code to put it all together is actually quite straightforward. The Argo ”runtime” code, which is used to construct the sandbox your extension will run in, looks almost exactly like this for all Argo-capable surface areas in Shopify:

Reviewer 3: ```suggestion
While remote-ui provides the component model, and takes care of propagating updates to an extension’s UI into the host application, it does not have any built-in notion of an “extension” at all. To create an extension system, Shopify authors a bit of extra code: a mapping of “extension points” (which, in code, are just strings with a specific naming format), and a way to load third-party code that can register to be called for those extension points. This might sound little complicated, but the code to put it all together is actually quite straightforward. The Argo ”runtime” code, which is used to construct the sandbox your extension will run in, looks almost exactly like this for all Argo-capable surface areas in Shopify:
```

```js
// We keep a mapping of all the extension points you register for
const registeredExtensionPoints = new Map();

// We define a globally-available `shopify` object. The only thing this object
// can do is register an extension (which is just a callback function) for a
// specific extension point.

> globalThis.shopify = {

Reviewer 4: `self=globalThis` so this shouldn't matter. However, we're currently calling this `self` in all of our libraries so we should make it consistent in the doc

  extend(extensionPoint, extension) {
    registeredExtensionPoints.set(extensionPoint, extension);
  },
};

// Here’s the function the host application will call to load your extension’s
// script into the sandbox
export function load(script) {
  // Internal workings — don’t worry too much about this, because we might change

>   // it in the future as new web platform features are available!

Reviewer 4: I think you're talking about `reload` @henrytao-me. I think this is supposed to be the function called by the host to import the external script

>   // it in the future as new web platform features are available!

Reviewer 5: 👍

}

> // Once your script is loaded, it has had its chance to register for extension

Reviewer 4: ```suggestion
// Once your script is loaded, it can register callbacks for extension points.
// Executing this function results in calling your registered callbacks with arguments from
// the host application, like the data your extension has access to and the UI “root“
// your components will be attached to.
```

// points, and we can now simply call your extension with the arguments from the
// host application, like the data your extension has access to and the UI “root“

> // your components will be attached to.

Reviewer 6: I'm having some trouble understanding this sentence.

export function run(extensionPoint, ...args) {
  return registeredExtensionPoints.get(extensionPoint)?.(...args);
}
```

In your extension code, you make use of these APIs using patterns that feel idiomatic in JavaScript:

```js
shopify.extend('Checkout::Feature::Render', (root) => {
  root.appendChild(root.createText('Hello world!'));
});

// Or, if you are using our NPM libraries, we offer a bit of sugar over the
// global:

> import {extend} from '@shopify/argo-checkout';

Reviewer 4: Should we add a React example to illustrate how it works or is it too complicated?

Author: I'd prefer to avoid the framework wrappers in here, since we're mostly discussing the ”bare metal” of Argo.

extend('Checkout::Feature::Render', (root) => {
  root.appendChild(root.createText('Hello world!'));
});

> ```

Reviewer 6: It might be more helpful to put this code first as it illustrates the interface. The implementation model would make more sense following that.

Author: Hmm, I think that flow is kind of confusing, because we go from talking about the implementation in previous sections, to talking about extension code, to talking about the implementation again. I intended for the implementation of Argo to be the focus here; there's plenty of discussion that entirely focuses on the consumer code in the actual library documentation.

> As with UI components, the Argo extensibility framework lets us provide different arguments to your extension, depending the extension point you selected and other factors, like your app’s permissions. For full details on what arguments are passed for each extension point, please refer to the documentation for the [Argo-supporting surface area you want to extend](#i-just-want-to-build-an-argo-extension-not-learn-about-them).

Reviewer 6: typo: ...different arguments to your extension, depending **on** the extension point...

  return registeredExtensionPoints.get(extensionPoint)?.(...args);
}
```

In your extension code, you make use of these APIs using patterns that feel idiomatic in JavaScript:

```js
shopify.extend('Checkout::Feature::Render', (root) => {
  root.appendChild(root.createText('Hello world!'));
});

// Or, if you are using our NPM libraries, we offer a bit of sugar over the
// global:

import {extend} from '@shopify/argo-checkout';

extend('Checkout::Feature::Render', (root) => {
  root.appendChild(root.createText('Hello world!'));
});
```

As with UI components, the Argo extensibility framework lets us provide different arguments to your extension, depending the extension point you selected and other factors, like your app’s permissions. For full details on what arguments are passed for each extension point, please refer to the documentation for the [Argo-supporting surface area you want to extend](#i-just-want-to-build-an-argo-extension-not-learn-about-them).

## Contributing

> Check out our [contribution guidelines](/CONTRIBUTING.md).

Author: Good to know, I will add it back with a copy of these notes 👍

> We provide the libraries in this repo as public NPM packages so that they can be installed and used in your local development. Because the packages are public, we have also made the repo public, so you can more easily refer to the in-code comments we make heavy use of for documenting Argo APIs. However, the code in this repo is not a traditional open source project.

Reviewer 3: ```suggestion
We provide the libraries in this repo as public NPM packages so that they can be installed and used in your local project. Because the packages are public, we have also made the repo public. This way, you can easily refer to the in-code comments we use for documenting Argo APIs. However, the code in this repo is not a traditional open source project.
```

> Though we are not accepting contributions that, we’d still love to hear from you! If you have ideas for new components or APIs, please [raise an issue on this repo](https://github.com/Shopify/argo/issues/new/choose). We will also happily accept pull requests for fixing typos in the documentation. If you do raise an issue or PR on this repo, please read [the code of conduct](./CODE_OF_CONDUCT.md), which all contributors must adhere to.

Reviewer 6: Typo in the first part of this sentence (maybe remove _that_?)
